accepted/2021/experimental-features/experimental-features.md

```diff
@@ -0,0 +1,380 @@
+  using experimental features.
+* Customers can't accidentally depend on experimental features, either directly
+  or indirectly through some library.
+* We can deliver a reasonable experience even when an experimental feature can't
+  be supported in all the workloads, for example, by failing with a sensible
+  error message.
+* We have a unified experience for turning on experimental features, no matter
+  whether they are in a runtime, a compiler, a library, or span multiple parts
+  of the system.
+* We don't have to do unnatural acts to add an experimental feature, by, for
+  example, having to build a plug-in based system that allows swapping in and
+  out files.
+
+### Non-Goals
+
+* Allowing Microsoft to turn features on or off remotely, e.g. A/B testing
+* Designing a mechanism for shipping experimental features as separate packages
+  or installers
+
+## Stakeholders and Reviewers
+
+* Runtime teams
+* C#, VB, F# teams
+* Library teams
+* SDK team
+* MSBuild team
+* NuGet team
+
+## Design
+
+### Property in project files
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <ExperimentalFeatures>StaticInterfaceMembers</ExperimentalFeatures>
+  </PropertyGroup>
+
+</Project>
+```
+
+### Assembly info generation
+
+When we generate the `AssemblyInfo.cs`, we'll include all experimental features
+as an assembly level attribute (the attribute and the meaning is discussed
+[later in this document](#api-analyzer)).
+
+For example, this project information:
+
+```xml
+<Project>
+
+  <PropertyGroup>
+    <ExperimentalFeatures>StaticInterfaceMembers;FancyGC</ExperimentalFeatures>
+  </PropertyGroup>
+
+</Project>
+```
+
+Would result in these assembly-level attributes:
+
+```C#
+[assembly: ExperimentalFeature("StaticInterfaceMembers")]
+[assembly: ExperimentalFeature("FancyGC")]
+```
+
+This means you can look at a binary and tell which experimental features were
+used to build it. This can be used for auditing and for detecting transitive
+dependencies ([discussed later](#api-analyzer)).
+
+### runtime.json
+
+The `ExperimentalFeatures` should be written into the `runtime.json` file such
+that the host/runtime can use it to make decisions.
+
+### Compilation context
+
+The `ExperimentalFeatures` needs to be passed to the compilers. Roslyn (C#/VB)
+already has a mechanism for this which is exposed to both the compiler as well
+as for analyzers/source generators. Alternatively, we could say the information
+about experimental features can be retrieved from the generated assembly-level
+attributes.
+
+***Note:*** F# doesn't have this capability today.
+
+### API Analyzer
+
+We'll add an attribute that we can put on APIs we consider experimental. In some
+cases it might not be possible to mark an API (for example, because it's a new
+type system feature) in which case the feature would either be guarded by new
+language syntax or by a custom analyzer.
```

Comment:
The current design doesn't seem to protect against cases like this. Let's assume for a minute that the runtime team has done the work to enable `ExperimentalFeatures` for static interfaces then someone defines the following API in the system. 

```c#
public void SubtleMethod<T>(T t) where T : ISomeName { 
  T.SomeMethod();
}
```

It's easy at a glance to miss that this is in fact using static interface members. Yet nothing I see in this doc would offer a back stop to enforce that this API was indeed marked with `ExperimentalFeature`. We're essentially depending on the diligence of code reviews. 

Yes on the consumption side the compiler would not let you consume `SubtleMethod` unless you enabled experimental features (compiler would catch the static interface usage). But at the same time nothing would stop the runtime from shipping an API with RTM expectations (lack of `ExperimentalFeature` implies it's a real API) here. 

What happens when we rip this out in later releases?

Reply:
It's called out in the section you're standing on:

> We'll add an attribute that we can put on APIs we consider experimental. In some cases it might not be possible to mark an API (for example, because it's a new type system feature) in which case the feature would either be guarded by new language syntax or **by a custom analyzer**.

But I would generally assume that if we apply the attribute to an interface, we'd flag implementations and and interface constraints as well.

Reply:
I think you're missing my point here. My point isn't that we lack mechanisms to enforce this pattern when we see it. My point is there are no mechanisms in place to enforce we add these mechanisms where needed. Basically any one could come along, put this into a PR and unless a human was there to spot it lacked the attribute nothing else would stop it from being merged. Then we'd be left with an experimental API that wasn't marked as such.